### ISSUE TITLE
How do I use GTest's gtest_discover_tests()?

### ISSUE THREAD

**Issue by gc435** [microsoft/vcpkg#14853]
The [instructions in the vcpkg port](https://github.com/microsoft/vcpkg/blob/master/ports/gtest/usage) for GTest mostly work for me.  I had to make one change:

From:

```
target_link_libraries(main PRIVATE GTest::gtest GTest::gtest_main GTest::gmock GTest::gmock_main)
```

To:

```
target_link_libraries(main PRIVATE GTest::gtest GTest::gtest_main)
```

Without that change, no tests are found.

If I understand Craig Scott's [Professional CMake](https://crascit.com/professional-cmake/) and the [GoogleTest CMake documentation](https://cmake.org/cmake/help/git-stage/module/GoogleTest.html), I can use two new commands if I do `include(GoogleTest)` instead of `find_package(GTest CONFIG REQUIRED)`.  The new commands are `gtest_add_tests()` and `gtest_discover_tests()`.  However, when I use `include(GoogleTest)`, the compiler fails on `#include <gtest/gtest.h>`.  I don't have that problem with `find_package(GTest CONFIG REQUIRED)`.

I mostly use MSVC 2019.

Questions:

1. Does the above change to `target_link_libraries()` make sense?
2. Why does `include(GoogleTest)` not result in the preprocessor finding `<gtest/gtest.h>`?
3. When installed with `vcpkg`, will GTest be built from source when I compile my project?  Or, are the installed GTest `*.lib` libraries linked to without recompiling?  I see GTest DLLs in my build directory, but I'm unsure where they come from.  I also don't fully understand libraries on Windows.

Where I looked for an answer:

1. https://github.com/microsoft/vcpkg/blob/master/ports/gtest/usage
2. [Professional CMake](https://crascit.com/professional-cmake/) by Craig Scott
3. [GoogleTest documentation](https://cmake.org/cmake/help/git-stage/module/GoogleTest.html)

**Comment by JackBoosY** [comment]
Yeah, the message only tells you what are the targets you can use, not that you must use them.
And these targes are to use cmake configuration files generated by gtest(In _VCPKG_ROOT/installed/TRIPLET/share/gtest_), not `FindGTest.cmake` officially provided by cmake, so the documentation you provide is invalid.
When installing gtest, vcpkg will completely build it, then install it's headers, libraries, configuration files and some docs to the installed path.

**Comment by gc435** [comment]
Thanks for the thoughts, @JackBoosY.

> And these targes are to use cmake configuration files generated by gtest(In VCPKG_ROOT/installed/TRIPLET/share/gtest), not FindGTest.cmake officially provided by cmake, so the documentation you provide is invalid.

I'm still confused how to use `gtest_discover_tests()`.  It sounds like I shoud not use `find_package(GTest CONFIG REQUIRED)`.  Use `include(GoogleTest)`, right?

Also, any thoughts about this?

_2. Why does include(GoogleTest) not result in the preprocessor finding <gtest/gtest.h>?_

**Comment by JackBoosY** [comment]
As I said, Since we use CONFIG mode instead of MODULE mode, we don’t need to use the [GoogleTest.cmake](https://github.com/Kitware/CMake/blob/master/Modules/GoogleTest.cmake) provided by cmake.

**Comment by gc435** [comment]
> As I said, Since we use CONFIG mode instead of MODULE mode, we don’t need to use the GoogleTest.cmake provided by cmake.

I am not a CMake expert, so I don't understand the implications of this.  I now have a little better understanding of how GTest in `vcpkg` is different from GTest installed in the conventional way.  However, I still don't know to use `gtest_discover_tests()`.  Is it possible?

**Comment by JackBoosY** [comment]
See [documentation](https://cmake.org/cmake/help/latest/command/find_package.html).
If you always need to use `gtest_discover_tests`,  please use module mode `find_package(GTest REQUIRED)` and `include(GoogleTest)`. Because it's provided by cmake.

**Comment by ThomsonTan** [comment]
@JackBoosY as GoogleTest.cmake is a CMake built-in module, can I use it together with the gtest installation by vcpkg? So use both `find_package(GTest CONFIG REQUIRED)` and `include(GoogleTest)` at the same time and use `gtest_discover_tests`?

**Comment by JackBoosY** [comment]
@ThomsonTan  That's fine.

**Comment by craigscott-crascit** [comment]
Let me clarify some things, since there appear to have been some erroneous claims so far. You should be able to use `include(GoogleTest)` independently of how you have made GTest available to the build. Whether you use CMake's built-in `FindGTest` module, use `find_package()` to obtain it from vcpkg or some existing pre-built location, or even building GTest directly as part of the main build (see the [FetchContent](https://cmake.org/cmake/help/latest/module/FetchContent.html#examples) module for an example of this method), you should still be able to include the `GoogleTest` module and use the commands it provides.

> Why does include(GoogleTest) not result in the preprocessor finding <gtest/gtest.h>?

That isn't the job of `include(GoogleTest)`. If you are linking against the `gtest::gtest` target, it should be setting up the header search path for you. So whatever is creating that `gtest::gtest` target, that's where you'd need to look. Earlier versions of gtest didn't do this, but later versions do. I don't know if vcpkg is faithfully preserving that or not, but it is vcpkg's job to get this right, not the `GoogleTest` module.

> And these targes are to use cmake configuration files generated by gtest(In VCPKG_ROOT/installed/TRIPLET/share/gtest), not FindGTest.cmake officially provided by cmake, so the documentation you provide is invalid. When installing gtest, vcpkg will completely build it, then install it's headers, libraries, configuration files and some docs to the installed path.

CMake's `FindGTest` module first looks for a config file, so if vcpkg is creating one and the main build has been set up with appropriate search paths (`CMAKE_PREFIX_PATH`, etc.), then `FindGTest` should find it just fine.

> I'm still confused how to use gtest_discover_tests(). It sounds like I shoud not use find_package(GTest CONFIG REQUIRED). Use include(GoogleTest), right?
>
> Also, any thoughts about this?

The two things serve different purposes. A call to `find_package()` is about adding the gtest targets to your build. The `include(GoogleTest)` call is about defining the `gtest_add_test()` and `gtest_discover_tests()` commands so your project can use them. The two are only related by the fact that CMake's `FindGTest` module will automatically include the `GoogleTest` module for you, whereas a `find_package(GTest)` might not (it depends on what provides the package). As I stated above, you should always be able to call `include(GoogleTest)` regardless of how you are adding gtest to your build.

> If you always need to use gtest_discover_tests, please use module mode find_package(GTest REQUIRED) and include(GoogleTest).

This is partly incorrect. Please ignore the advice to use module mode `find_package(GTest REQUIRED)`, but do follow the advice to always call `include(GoogleTest)`.

**Comment by aboelens** [comment]
I have the same issue when linking against gmock using:

`target_link_libraries(main PRIVATE GTest::gtest GTest::gtest_main GTest::gmock GTest::gmock_main)`

that no test are detected using` gtest_discover_tests()`.

As a workaround I compile my tests with a file main.cpp which contains:

```
#include "gmock/gmock.h"
#include "gtest/gtest.h"

int main(int argc, char *argv[]) {
  ::testing::InitGoogleTest(&argc, argv);

  return RUN_ALL_TESTS();
}
```
